### PR TITLE
Tests and core rules

### PR DESCRIPTION
**Pull request by alvp**

### CODE REVIEW

File: src/jollyjumper/core.py [hunk 1]
```diff
@@ -12,26 +11,40 @@ def get_enjambment(text):
...
     """
-    enjambment_types = ['sirrematic']
+    enjambment_types = ['sirrematic', 'sirrematic_relation_words_preposition',
+                        'sirrematic_relation_words_conjunction',
+                        'sirrematic_relation_words_determiners',
+                        'sirrematic_orational', 'relation_words_verbs',
+                        'sirrematic_with_verb', 'link',
+                        'sirrematic_prepositional_without_de',
+                        'sirrematic_prepositional',
+                        'sirrematic_prepositional_prep_before_noun_or_adjective']
```
Comment (versae) [inline src/jollyjumper/core.py]: This is now `sirrematic_prepositional_before_noun_adjective`.

File: setup.py [hunk 2]
```diff
@@ -52,7 +58,7 @@ def read(*names, **kwargs):
     include_package_data=True,
     zip_safe=False,
     classifiers=[
-        # complete classifier list: http://pypi.python.org/pypi?%3Aaction=list_classifiers
+        # complete class list: http://pypi.python.org/pypi?%3Aaction=list_classifiers
```
Comment (versae) [inline setup.py]: This should be left as `classifier`

File: src/jollyjumper/core.py [hunk 3]
```diff
@@ -12,26 +11,40 @@ def get_enjambment(text):
...
-        previous_token = doc[token.i-1]
+        previous_token = doc[token.i - 1]
         next_token = doc[token.i + 1]
-        # We look for enjambment when if there are words before and after a newline character
-        if token.text == '\n' and not previous_token.is_punct and not next_token.is_punct:
+        # We look for enjambment when if there are words
+        # before and after a newline character
+        if token.text == '\n' \
+                and not previous_token.is_punct and not next_token.is_punct:
             for enjambment_type in enjambment_types:
```
Comment (versae) [inline src/jollyjumper/core.py:40]: Let's try to avoid using `\` and instead use parenthesis and indentation.

File: src/jollyjumper/core.py [hunk 4]
```diff
@@ -12,26 +11,40 @@ def get_enjambment(text):
...
-                enjambment_func = rules.get(f'get_{enjambment_type}_enjambment', lambda *_: None)
+                enjambment_func = rules.get(f'get_{enjambment_type}_enjambment',
+                                            lambda *_: None)
                 enjambment = enjambment_func(previous_token, next_token)
                 if enjambment:
-                    enjambments[token._.line] = (enjambment_type, enjambment)  # noqa
+                    enjambments[token._.line] = (
+                    enjambment_type, enjambment)  # noqa
```
Comment (versae) [inline src/jollyjumper/core.py]: Check indentation, maybe?

File: src/jollyjumper/rules.py [hunk 5]
```diff
@@ -1,3 +1,13 @@
+import re
+
+ENJAMBMENT_PREPOSITIONS = (
+    "a", "al", "ante", "bajo", "con", "contra", "desde", "en", "entre", "hacia",
+    "hasta", "para",
+    "por", "según", "sin", "sobre", "tras", "mediante", "durante", "salvo",
+    "excepto", "cabe",
+    "so")
+
```
Comment (versae) [inline src/jollyjumper/rules.py:6]: Check indentation.

File: src/jollyjumper/rules.py [hunk 6]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
                          ]
     while sirremactic_pairs:
         sirrematic_pair = sirremactic_pairs.pop()
-        if sorted((previous_token.pos_, next_token.pos_)) == sorted(sirrematic_pair)\
-                and (next_token.is_ancestor(previous_token) or previous_token.is_ancestor(next_token)):
+        if sorted((previous_token.pos_, next_token.pos_)) == sorted(
+                sirrematic_pair) \
+                and (next_token.is_ancestor(previous_token)
+                     or previous_token.is_ancestor(next_token)):
```
Comment (versae) [inline src/jollyjumper/rules.py:27]: Try using parenthesis instead.

File: src/jollyjumper/rules.py [hunk 7]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
\ No newline at end of file
+def get_sirrematic_relation_words_preposition_enjambment(previous_token,
+                                                         next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if previous_token.lower_ in ENJAMBMENT_PREPOSITIONS and re.search(
+            'AdpType=Prep', previous_token.tag_):
```
Comment (versae) [inline src/jollyjumper/rules.py:40]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 8]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+        return [previous_token.pos_, next_token.pos_]
+    return None
```
Comment (versae) [inline src/jollyjumper/rules.py:42]: Try using `return previous_token.pos_, next_token.pos_` and replace where needed.

File: src/jollyjumper/rules.py [hunk 9]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+def get_sirrematic_relation_words_determiners_enjambment(previous_token,
+                                                         next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if previous_token.pos_ == 'DET' and next_token.pos_ in (
+            'NOUN', 'ADJ', 'ADV', 'DET'):
```
Comment (versae) [inline src/jollyjumper/rules.py:66]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 10]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+def get_sirrematic_relation_words_verbs_enjambment(previous_token, next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if previous_token.pos_ in ('AUX', 'VERB') and re.search('AdpType=Prep',
+                                                            next_token.tag_) \
+            and next_token.lower_ in ('de', 'del'):
```
Comment (versae) [inline src/jollyjumper/rules.py:92]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 11]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+
+def get_sirrematic_orational_enjambment(previous_token, next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if previous_token.pos_ in ('ADJ', 'NOUN', 'ADV') \
+            and re.search('NumType', next_token.tag_) \
```
Comment (versae) [inline src/jollyjumper/rules.py:105]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 12]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+def get_sirrematic_prepositional_enjambment(previous_token, next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if next_token.n_rights > 0:
+        if previous_token.pos_ in ('ADJ', 'ADV', 'NOUN') and re.search(
+                'AdpType=Prep', next_token.tag_) \
```
Comment (versae) [inline src/jollyjumper/rules.py:121]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 13]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+                                                       next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if next_token.n_rights > 0:
+        if previous_token.pos_ in ('ADJ', 'NOUN') \
+                and re.search('AdpType=Prep', next_token.tag_) \
```
Comment (versae) [inline src/jollyjumper/rules.py:137]: Maybe better between parenthesis. Plus, the first two `if`s could be merged into one.

File: src/jollyjumper/rules.py [hunk 14]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+        previous_token, next_token):
+    """
+    Checks if sirrematic enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Sirrematic type or None if not found
+    """
+    if next_token.n_rights > 0:
+        if previous_token.dep_ in (
+                'ROOT', 'nsubj') and previous_token.pos_ == 'NOUN' \
```
Comment (versae) [inline src/jollyjumper/rules.py:154]: Maybe better between parenthesis.

File: src/jollyjumper/rules.py [hunk 15]
```diff
@@ -14,10 +24,158 @@ def get_sirrematic_enjambment(previous_token, next_token):
...
+
+def get_link_enjambment(previous_token, next_token):
+    """
+    Checks if a link enjambment exists between two lines
+    :param previous_token: The word before a newline character
+    :param next_token: The word before a newline character
+    :return: Link type or None if not found
+    """
+    if next_token.n_rights > 0:
+        if previous_token.dep_ in ('ROOT', 'nsubj') \
```
Comment (versae) [inline src/jollyjumper/rules.py:170]: Maybe better between parenthesis.

File: tests/test_core.py [hunk 16]
```diff
@@ -1 +1,79 @@
...
+            TokenMock(n_rights=1, tag_="NumType", pos_="ADJ", text="mi", i=0,
+                      is_punct=False, has_tmesis=False,
+                      line=0),
+            TokenMock(n_rights=1, tag_="NumType", pos_="SPACE", text="\n", i=1,
+                      is_punct=False, has_tmesis=False,
+                      line=0),
+            TokenMock(n_rights=1, tag_="NumType", pos_="NOUN", text="casa", i=2,
+                      is_punct=False, has_tmesis=False,
+                      line=1)
+        ]
```
Comment (versae) [inline tests/test_core.py:56]: Check line numbers.

File: tests/test_core.py [hunk 17]
```diff
@@ -1 +1,79 @@
...
+        return lambda _: [
+            TokenMock(n_rights=1, tag_="NumType", pos_="ADJ", text="mi-", i=0,
+                      is_punct=False, has_tmesis=False,
+                      line=1),
+            TokenMock(n_rights=1, tag_="NumType", pos_="SPACE", text="\n", i=0,
+                      is_punct=False, has_tmesis=False,
+                      line=1),
+            TokenMock(n_rights=1, tag_="NumType", pos_="NOUN", text="ro", i=0,
+                      is_punct=False, has_tmesis=False,
+                      line=2)
```
Comment (versae) [inline tests/test_core.py:74]: Check line numbers.

File: tests/test_pipeline.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+test_dict_list = [
+    {'text': 'esto', 'pos_': '', 'tag_': '', '_.line': 0, '_.has_tmesis': False,
+     'lower_': 'esto', 'n_rights': 0},
+    {'text': 'es', 'pos_': '', 'tag_': '', '_.line': 0, '_.has_tmesis': False,
+     'lower_': 'es', 'n_rights': 0},
+    {'text': 'una', 'pos_': '', 'tag_': '', '_.line': 0, '_.has_tmesis': False,
+     'lower_': 'una', 'n_rights': 0},
+    {'text': 'prue-\nba', 'pos_': '', 'tag_': '', '_.line': 0,
+     '_.has_tmesis': True, 'lower_': 'prue-\nba',
+     'n_rights': 0}]
```
Comment (versae) [inline tests/test_pipeline.py:14]: Are line numbers right here?

File: tests/test_rules.py [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+    assert get_sirrematic_enjambment(previous_token, next_token) == ['ADV',
+                                                                     'VERB']
+
+
+def test_get_sirrematic_relation_words_preposition_enjambment_adp_dummy():
+    previous_token = TokenMock(lower_="a", pos_="ADP", tag_="AdpType=Prep")
+    next_token = TokenMock(pos_="dummy")
+    assert get_sirrematic_relation_words_preposition_enjambment(previous_token,
+                                                                next_token) == [
+               'ADP', 'dummy']
```
Comment (versae) [inline tests/test_rules.py]: Looks weird.

File: tests/test_rules.py [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+
+
+def test_get_sirrematic_relation_words_conjunction_enjambment_conj_dummy():
+    previous_token = TokenMock(lower_="a", pos_="CONJ", tag_="AdpType=Prep")
+    next_token = TokenMock(pos_="dummy")
+    assert get_sirrematic_relation_words_conjunction_enjambment(previous_token,
+                                                                next_token) == [
+               'CONJ', 'dummy']
```
Comment (versae) [inline tests/test_rules.py]: Looks weird.

File: tests/test_rules.py [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+
+
+def test_get_sirrematic_relation_words_determiners_enjambment_det_noun():
+    previous_token = TokenMock(pos_="DET")
+    next_token = TokenMock(pos_="NOUN")
+    assert get_sirrematic_relation_words_determiners_enjambment(previous_token,
+                                                                next_token) == [
+               'DET', 'NOUN']
```
Comment (versae) [inline tests/test_rules.py]: Looks weird.

File: tox.ini [hunk 22]
```diff
@@ -23,7 +23,7 @@ commands =
 
 [coverage:run]
 omit =
-    src/skas/__main__.py
+    src/jollyjumper/__main__.py
```
Comment (versae) [inline tox.ini:26]: :+1:

**Review by versae — CHANGES_REQUESTED** [review]
Mostly cosmetic issues.

File: src/jollyjumper/core.py (hunk 1 above)
Comment (alvp) [reply src/jollyjumper/core.py]: Done

File: setup.py (hunk 2 above)
Comment (alvp) [reply setup.py]: Done

File: src/jollyjumper/core.py (hunk 3 above)
Comment (alvp) [reply src/jollyjumper/core.py:40]: Done

File: src/jollyjumper/core.py (hunk 4 above)
Comment (alvp) [reply src/jollyjumper/core.py]: Done

File: src/jollyjumper/rules.py (hunk 5 above)
Comment (alvp) [reply src/jollyjumper/rules.py:6]: Done

File: tests/test_pipeline.py (hunk 18 above)
Comment (alvp) [reply tests/test_pipeline.py:14]: Ignored, works as intended

File: tests/test_rules.py (hunk 19 above)
Comment (alvp) [reply tests/test_rules.py]: Done

File: tests/test_rules.py (hunk 20 above)
Comment (alvp) [reply tests/test_rules.py]: SO done

File: tests/test_rules.py (hunk 21 above)
Comment (alvp) [reply tests/test_rules.py]: SO very done

**Review by versae — APPROVED** [review]

**Pull request merged**